In [1]:

#IMPORTS
from docx import Document
import pandas as pd

In [2]:
docx_file_path = 'NewFacultyOrientationResearchFAQs2021.docx'
doc = Document(docx_file_path)

In [3]:
# Create a text file to write the extracted text, this copies the text from the doc to txt
txt_file_path = 'EditedNewFacultyOrientationResearch.txt'
with open(txt_file_path, 'w', encoding='utf-8') as txt_file:
    for paragraph in doc.paragraphs:
        txt_file.write(paragraph.text + '\n')

print(f'Text extracted from {docx_file_path} and saved to {txt_file_path}')


Text extracted from NewFacultyOrientationResearchFAQs2021.docx and saved to EditedNewFacultyOrientationResearch.txt


In [4]:
#Local file path
txt_file_path = '/Users/avinash/EditedNewFacultyOrientationResearch.txt'


In [5]:
#making the lists for the different entries
questions = []
answers = []
tags = []

# Read the text file and iterate through non-empty lines
with open(txt_file_path, 'r', encoding='utf-8') as txt_file:
    lines = [line.strip() for line in txt_file.readlines() if line.strip()]
    
    # Iterate through lines in sets of 3
    for i in range(0, len(lines), 3):
        # Check if there are enough lines left to extract a set
        # currently dependent on the three way format
        if i + 2 < len(lines):
            questions.append(lines[i])
            answers.append(lines[i + 1])
            tags.append(lines[i + 2])
        else:
            # Not enough lines, skip
            print(f"Skipping incomplete set at index {i}.")

# dataFrame from the lists of questions, answers, and tags
df = pd.DataFrame({'Question': questions, 'Answer': answers, 'Tag': tags})
df.head()

Question  \
0  I would like to submit a proposal, how do I ge...   
1  Please start the first page, General tab, of t...   
2  How will I know who my assigned research admin...   
3  No. OSRP handles all sponsored projects for th...   
4  I am working with an industrial partner, and t...   

                                              Answer  \
0  Please start a routing sheet (SP proposal) in ...   
1                 What are IIT’s internal deadlines?   
2  After you start your routing sheet, OSRP will ...   
3  As a faculty member, do I have authority to le...   
4  Please contact osrp@iit.edu. OSRP will assist ...   

                                                 Tag  
0   How soon should I start my Cayuse routing sheet?  
1  The administrative portions of the proposal (b...  
2  OSRP only handles federally sponsored projects...  
3  No. Faculty do not have the authority to legal...  
4  What are indirect costs (also called facilitie...

In [11]:
# Read the Word document and create a text file
doc = Document(docx_file_path)
with open(txt_file_path, 'w', encoding='utf-8') as txt_file:
    for paragraph in doc.paragraphs:
        txt_file.write(paragraph.text + '\n')


# Function to assign tags based on keywords
def assign_tags(question):
    if 'proposal' in question.lower():
        return 'proposal'
    elif 'routing sheet' in question.lower():
        return 'routing sheet'
    elif 'deadlines' in question.lower():
        return 'deadlines'
    elif 'research admin' in question.lower():
        return 'research admin'
    elif 'sponsored projects' in question.lower():
        return 'sponsored projects'
    elif 'bind' in question.lower():
        return 'bind'
    elif 'non-disclosure agreement (NDA)' in question.lower():
        return 'non-disclosure agreement (NDA)'
    else:
        return 'other'
    
# Lists for different entries
questions = []
answers = []
tags = []

# Read the text file and iterate through non-empty lines
with open(txt_file_path, 'r', encoding='utf-8') as txt_file:
    lines = [line.strip() for line in txt_file.readlines() if line.strip()]

    # Iterate through lines in sets of 3
    for i in range(0, len(lines), 3):
        # Check if there are enough lines left to extract a set
        if i + 2 < len(lines):
            question = lines[i]
            questions.append(question)
            answers.append(lines[i + 1])
            
            # Assign tags based on the function
            tags.append(assign_tags(question))
        else:
            # Not enough lines, skip
            print(f"Skipping incomplete set at index {i}.")

# DataFrame from the lists of questions, answers, and tags
df = pd.DataFrame({'Question': questions, 'Answer': answers, 'Tag': tags})

# Display the DataFrame
df.head()


Question  \
0  I would like to submit a proposal, how do I ge...   
1  Please start the first page, General tab, of t...   
2  How will I know who my assigned research admin...   
3  No. OSRP handles all sponsored projects for th...   
4  I am working with an industrial partner, and t...   

                                              Answer                 Tag  
0  Please start a routing sheet (SP proposal) in ...            proposal  
1                 What are IIT’s internal deadlines?            proposal  
2  After you start your routing sheet, OSRP will ...      research admin  
3  As a faculty member, do I have authority to le...  sponsored projects  
4  Please contact osrp@iit.edu. OSRP will assist ...               other

In [12]:
#copying the dataframe, so I do not mess up the first one
df_copy = df.copy()


In [14]:
# Adding another column of similar questions to the dataframe
new_column_values = [
    "How do I initiate the process of submitting a proposal?",
    "When should I begin filling out my Cayuse routing sheet?",
    "What are the internal deadlines set by IIT?",
    "How can I find out the identity of my assigned research administrator?",
    "Is it accurate to say that OSRP exclusively handles federally sponsored projects?",
    "As a faculty member, am I authorized to enter into legally binding agreements on behalf of Illinois Tech?",
    "I've been requested by an industrial partner to sign a non-disclosure agreement (NDA). What steps should I take?",
    "Could you explain what indirect costs (also known as facilities and administrative or F&A costs) entail?",
    "What are the indirect cost rates applicable to Illinois Tech?",
    
]

# Add the new column to the DataFrame
df['similarQ1'] = new_column_values
df.head()

Question  \
0  I would like to submit a proposal, how do I ge...   
1  Please start the first page, General tab, of t...   
2  How will I know who my assigned research admin...   
3  No. OSRP handles all sponsored projects for th...   
4  I am working with an industrial partner, and t...   

                                              Answer                 Tag  \
0  Please start a routing sheet (SP proposal) in ...            proposal   
1                 What are IIT’s internal deadlines?            proposal   
2  After you start your routing sheet, OSRP will ...      research admin   
3  As a faculty member, do I have authority to le...  sponsored projects   
4  Please contact osrp@iit.edu. OSRP will assist ...               other   

                                           similarQ1  
0  How do I initiate the process of submitting a ...  
1  When should I begin filling out my Cayuse rout...  
2        What are the internal deadlines set by IIT?  
3  How can I find out the identity of my assigned...  
4  Is it accurate to say that OSRP exclusively ha...

In [15]:
new_column_values = [
    "How do I begin the process of drafting and submitting a proposal?",
    "When is the optimal time to start completing my Cayuse routing sheet?",
    "What are the internal deadlines established by IIT that I need to be aware of?",
    "How can I determine the identity of the research administrator assigned to me?",
    "Is it correct to assume that OSRP exclusively manages projects funded by federal entities?",
    "As a faculty member, do I possess the authority to legally commit Illinois Tech to agreements?",
    "A corporate partner has asked me to sign a non-disclosure agreement (NDA). What are the recommended steps for handling this request?",
    "Can you provide an explanation of what is encompassed by indirect costs, also referred to as facilities and administrative or F&A costs?",
    "What are the prevailing indirect cost rates applicable to Illinois Tech's sponsored projects?",
    ]

# Add the new column to the DataFrame
df['similarQ2'] = new_column_values
df.head()

Question  \
0  I would like to submit a proposal, how do I ge...   
1  Please start the first page, General tab, of t...   
2  How will I know who my assigned research admin...   
3  No. OSRP handles all sponsored projects for th...   
4  I am working with an industrial partner, and t...   

                                              Answer                 Tag  \
0  Please start a routing sheet (SP proposal) in ...            proposal   
1                 What are IIT’s internal deadlines?            proposal   
2  After you start your routing sheet, OSRP will ...      research admin   
3  As a faculty member, do I have authority to le...  sponsored projects   
4  Please contact osrp@iit.edu. OSRP will assist ...               other   

                                           similarQ1  \
0  How do I initiate the process of submitting a ...   
1  When should I begin filling out my Cayuse rout...   
2        What are the internal deadlines set by IIT?   
3  How can I find out the identity of my assigned...   
4  Is it accurate to say that OSRP exclusively ha...   

                                           similarQ2  
0  How do I begin the process of drafting and sub...  
1  When is the optimal time to start completing m...  
2  What are the internal deadlines established by...  
3  How can I determine the identity of the resear...  
4  Is it correct to assume that OSRP exclusively ...

In [16]:
new_column_values = [
    "How can I get started with drafting and submitting a proposal?",
    "When should I aim to start completing my Cayuse routing sheet?",
    "What are the internal deadlines set by IIT that I should be mindful of?",
    "How do I find out who my assigned research administrator is?",
    "Does OSRP exclusively handle projects funded by federal sponsors?",
    "As a faculty member, am I empowered to legally commit Illinois Tech to agreements?",
    "I've been asked by a corporate partner to sign an NDA. What steps should I take?",
    "Can you explain the concept of indirect costs, also known as facilities and administrative or F&A costs?",
    "What are the current indirect cost rates applicable to sponsored projects at Illinois Tech?",
  
]

# Add the new column to the DataFrame
df['similarQ3'] = new_column_values
df.head()

Question  \
0  I would like to submit a proposal, how do I ge...   
1  Please start the first page, General tab, of t...   
2  How will I know who my assigned research admin...   
3  No. OSRP handles all sponsored projects for th...   
4  I am working with an industrial partner, and t...   

                                              Answer                 Tag  \
0  Please start a routing sheet (SP proposal) in ...            proposal   
1                 What are IIT’s internal deadlines?            proposal   
2  After you start your routing sheet, OSRP will ...      research admin   
3  As a faculty member, do I have authority to le...  sponsored projects   
4  Please contact osrp@iit.edu. OSRP will assist ...               other   

                                           similarQ1  \
0  How do I initiate the process of submitting a ...   
1  When should I begin filling out my Cayuse rout...   
2        What are the internal deadlines set by IIT?   
3  How can I find out the identity of my assigned...   
4  Is it accurate to say that OSRP exclusively ha...   

                                           similarQ2  \
0  How do I begin the process of drafting and sub...   
1  When is the optimal time to start completing m...   
2  What are the internal deadlines established by...   
3  How can I determine the identity of the resear...   
4  Is it correct to assume that OSRP exclusively ...   

                                           similarQ3  
0  How can I get started with drafting and submit...  
1  When should I aim to start completing my Cayus...  
2  What are the internal deadlines set by IIT tha...  
3  How do I find out who my assigned research adm...  
4  Does OSRP exclusively handle projects funded b...

In [20]:
# Embedding model
model = BGEM3FlagModel('BAAI/bge-m3', use_fp16=True)

# Sentences for embedding
questions_for_embedding = df['Question'].tolist()

# Embedding generation
question_embeddings = model.encode(questions_for_embedding)['dense_vecs']

# Adding embeddings to the DataFrame
df['Question_Embedding'] = list(question_embeddings)
# Provided similar questions
similar_questions_sets = [
    [
    "How do I initiate the process of submitting a proposal?",
    "When should I begin filling out my Cayuse routing sheet?",
    "What are the internal deadlines set by IIT?",
    "How can I find out the identity of my assigned research administrator?",
    "Is it accurate to say that OSRP exclusively handles federally sponsored projects?",
    "As a faculty member, am I authorized to enter into legally binding agreements on behalf of Illinois Tech?",
    "I've been requested by an industrial partner to sign a non-disclosure agreement (NDA). What steps should I take?",
    "Could you explain what indirect costs (also known as facilities and administrative or F&A costs) entail?",
    "What are the indirect cost rates applicable to Illinois Tech?",
    
],
    [
    "How do I begin the process of drafting and submitting a proposal?",
    "When is the optimal time to start completing my Cayuse routing sheet?",
    "What are the internal deadlines established by IIT that I need to be aware of?",
    "How can I determine the identity of the research administrator assigned to me?",
    "Is it correct to assume that OSRP exclusively manages projects funded by federal entities?",
    "As a faculty member, do I possess the authority to legally commit Illinois Tech to agreements?",
    "A corporate partner has asked me to sign a non-disclosure agreement (NDA). What are the recommended steps for handling this request?",
    "Can you provide an explanation of what is encompassed by indirect costs, also referred to as facilities and administrative or F&A costs?",
    "What are the prevailing indirect cost rates applicable to Illinois Tech's sponsored projects?",
    ],
    [
    "How can I get started with drafting and submitting a proposal?",
    "When should I aim to start completing my Cayuse routing sheet?",
    "What are the internal deadlines set by IIT that I should be mindful of?",
    "How do I find out who my assigned research administrator is?",
    "Does OSRP exclusively handle projects funded by federal sponsors?",
    "As a faculty member, am I empowered to legally commit Illinois Tech to agreements?",
    "I've been asked by a corporate partner to sign an NDA. What steps should I take?",
    "Can you explain the concept of indirect costs, also known as facilities and administrative or F&A costs?",
    "What are the current indirect cost rates applicable to sponsored projects at Illinois Tech?",
  
]
    
   
]

# Loop through each set of similar questions and calculate similarity
for i, similar_questions in enumerate(similar_questions_sets, start=1):
    similar_embeddings = model.encode(similar_questions)['dense_vecs']
    similarity = question_embeddings @ similar_embeddings.T
    df[f'Similarity_Set_{i}'] = list(similarity)

# Display the DataFrame with embeddings and similarity scores
df.head()


Fetching 21 files: 100%|██████████████████████████| 21/21 [00:00<00:00, 101475.10it/s]


loading existing colbert_linear and sparse_linear---------


encoding:   0%|                                                 | 0/1 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
encoding:   0%|                                                 | 0/1 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
encoding:   0%|                                                 | 0/1 [00:00<?, ?it/s]huggingface/tokenizers: The current process just got forked, after parallelism has already bee

Question  \
0  I would like to submit a proposal, how do I ge...   
1  Please start the first page, General tab, of t...   
2  How will I know who my assigned research admin...   
3  No. OSRP handles all sponsored projects for th...   
4  I am working with an industrial partner, and t...   

                                              Answer                 Tag  \
0  Please start a routing sheet (SP proposal) in ...            proposal   
1                 What are IIT’s internal deadlines?            proposal   
2  After you start your routing sheet, OSRP will ...      research admin   
3  As a faculty member, do I have authority to le...  sponsored projects   
4  Please contact osrp@iit.edu. OSRP will assist ...               other   

                                           similarQ1  \
0  How do I initiate the process of submitting a ...   
1  When should I begin filling out my Cayuse rout...   
2        What are the internal deadlines set by IIT?   
3  How can I find out the identity of my assigned...   
4  Is it accurate to say that OSRP exclusively ha...   

                                           similarQ2  \
0  How do I begin the process of drafting and sub...   
1  When is the optimal time to start completing m...   
2  What are the internal deadlines established by...   
3  How can I determine the identity of the resear...   
4  Is it correct to assume that OSRP exclusively ...   

                                           similarQ3  \
0  How can I get started with drafting and submit...   
1  When should I aim to start completing my Cayus...   
2  What are the internal deadlines set by IIT tha...   
3  How do I find out who my assigned research adm...   
4  Does OSRP exclusively handle projects funded b...   

                                  Question_Embedding  \
0  [-0.0656, 0.01817, -0.012054, 0.01226, -0.0074...   
1  [-0.02115, 0.01936, -0.01724, 0.01263, -0.0293...   
2  [-0.0707, -0.02513, -0.0425, -0.01508, -0.0207...   
3  [-0.01622, -0.0002224, -0.00836, -0.002668, -0...   
4  [-0.07446, 0.00889, -0.01953, 0.010254, -0.039...   

                                    Similarity_Set_1  \
0  [0.9614, 0.4854, 0.4058, 0.4644, 0.3193, 0.487...   
1  [0.5874, 0.74, 0.401, 0.3667, 0.4426, 0.3198, ...   
2  [0.4949, 0.3994, 0.441, 0.9688, 0.396, 0.465, ...   
3  [0.4092, 0.2927, 0.3098, 0.4548, 0.7324, 0.452...   
4  [0.541, 0.3647, 0.3745, 0.4883, 0.3237, 0.5, 0...   

                                    Similarity_Set_2  \
0  [0.924, 0.4834, 0.4724, 0.4807, 0.3174, 0.4856...   
1  [0.5825, 0.7324, 0.464, 0.3674, 0.4512, 0.313,...   
2  [0.4878, 0.396, 0.515, 0.948, 0.4194, 0.473, 0...   
3  [0.398, 0.2993, 0.3606, 0.455, 0.726, 0.4111, ...   
4  [0.5156, 0.358, 0.4644, 0.5, 0.3257, 0.475, 0....   

                                    Similarity_Set_3  
0  [0.933, 0.4968, 0.467, 0.4788, 0.3445, 0.4814,...  
1  [0.5776, 0.7324, 0.4568, 0.3892, 0.4355, 0.306...  
2  [0.4695, 0.3955, 0.4883, 0.9893, 0.405, 0.4644...  
3  [0.3916, 0.3066, 0.3396, 0.4722, 0.739, 0.415,...  
4  [0.4941, 0.3672, 0.452, 0.4875, 0.3325, 0.469,...

Similarity Set 1:

The first value 0.933 indicates the similarity score between the original question and the first question in Similarity Set 1.
The second value 0.4968 indicates the similarity score with the second question in the set, and so on.
Similarity Set 2:

Similar interpretation applies for Similarity Set 2.
Similarity Set 3:

Similar interpretation applies for Similarity Set 3.
General Interpretation:
Higher Similarity Scores:

Higher scores indicate a higher degree of similarity between the original question and the questions in the corresponding set.
Tag ("proposal"):

The assigned tag based on keyword matching in the original question.


First Value (0.933):

This value indicates the similarity score between the original question and the first question in the set. A score close to 1 suggests a high degree of similarity.
Second Value (0.4968):

Indicates the similarity score with the second question in the set.
Third Value (0.467):

Indicates the similarity score with the third question in the set.
Fourth Value (0.4788):

Indicates the similarity score with the fourth question in the set.
Fifth Value (0.3445):

Indicates the similarity score with the fifth question in the set.
Sixth Value (0.4814):

Indicates the similarity score with the sixth question in the set.